src/energy.cpp

```diff
@@ -164,166 +182,178 @@ void PolicyIonIon::updateComplex(EwaldData &data, Change &change, Space::Tgvec &
-            auto g_old = oldgroups.at(cg.index);
-            for (auto i : cg.atoms) {
-                if (i < g_new.size())
-                    Q += q.cwiseProduct((g_new.begin() + i)->pos).array().cos().prod() * (g_new.begin() + i)->charge;
-                if (i < g_old.size())
-                    Q -= q.cwiseProduct((g_old.begin() + i)->pos).array().cos().prod() * (g_old.begin() + i)->charge;
-            }
+    for (int k = 0; k < data.k_vectors.cols(); k++) {
+      auto &Q = data.Q_ion[k];
+      const Point &q = data.k_vectors.col(k);
+      for (auto &changed_group : change.groups) {
+        auto &g_new = groups.at(changed_group.index);
+        auto &g_old = oldgroups.at(changed_group.index);
+        for (auto i : changed_group.atoms) {
+          if (i < g_new.size())
+            Q += q.cwiseProduct(g_new[i].pos).array().cos().prod() *
+                 g_new[i].charge;
+          if (i < g_old.size())
+            Q -= q.cwiseProduct(g_old[i].pos).array().cos().prod() *
+                 g_old[i].charge;
         }
+      }
     }
 }
 
 double PolicyIonIon::surfaceEnergy(const EwaldData &d, Change &change, Space::Tgvec &groups) {
     if (d.const_inf < 0.5)
         return 0;
     Point qr(0, 0, 0);
-    if (change.all or change.dV)
-        for (auto g : groups)
-            for (auto i : g)
-                qr += i.charge * i.pos;
-    else if (change.groups.size() > 0) {
-        for (auto cg : change.groups) {
-            auto g = groups.at(cg.index);
-            for (auto i : cg.atoms)
-                if (i < g.size())
-                    qr += (g.begin() + i)->charge * (g.begin() + i)->pos;
+    if (change.all or change.dV) {
+      for (auto &g : groups) {
+        for (auto &particle : g) {
+          qr += particle.charge * particle.pos;
+        }
+      }
+    } else if (change.groups.size() > 0) {
+      for (auto &changed_group : change.groups) {
+        auto &g = groups.at(changed_group.index);
+        for (auto i : changed_group.atoms) {
+          if (i < g.size()) {
+            qr += g[i].charge * g[i].pos;
+          }
         }
+      }
     }
-    double volume = d.L.x() * d.L.y() * d.L.z();
+    double volume = d.box_length.prod();
     return d.const_inf * 2 * pc::pi / ((2 * d.surface_dielectric_constant + 1) * volume) * qr.dot(qr) *
            d.bjerrum_length;
 }
 
 double PolicyIonIon::selfEnergy(const EwaldData &d, Change &change, Space::Tgvec &groups) {
-    double Eq = 0;
-    if (change.dN) {
-        for (auto cg : change.groups) {
-            auto g = groups.at(cg.index);
-            for (auto i : cg.atoms)
-                if (i < g.size())
-                    Eq += std::pow((g.begin() + i)->charge, 2);
+  double energy = 0;
+  if (change.dN) {
+    for (auto &changed_group : change.groups) {
+      auto &g = groups.at(changed_group.index);
+      for (auto i : changed_group.atoms) {
+        if (i < g.size()) {
+          energy += std::pow((g.begin() + i)->charge, 2);
```

Comment:
Forgotten `(g.begin() + i)->…` to `g[i].…`

Reply:
fixed